cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,481 @@
+  explicit ColumnMerger(VectorI const& row_order,
+                        rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                        cudaStream_t stream = nullptr):
+    dv_row_order_(row_order),
+    mr_(mr),
+    stream_(stream)
+  {
+  }
+  
+  // column merger operator;
+  //
+  template<typename ElemenT>//required: column type
+  std::enable_if_t<cudf::is_fixed_width<ElemenT>(),
+                   std::unique_ptr<cudf::column>>
+  operator()(cudf::column_view const& lcol, cudf::column_view const& rcol) const
+  {
+    auto lsz = lcol.size();
+    auto merged_size = lsz + rcol.size();
+    auto type = lcol.type();
+    
+    std::unique_ptr<cudf::column> p_merged_col = cudf::experimental::allocate_like(lcol, merged_size);
+
+    //"gather" data from lcol, rcol according to dv_row_order_ "map"
+    //(directly calling gather() won't work because
+    // lcol, rcol indices overlap!)
+    //
+    cudf::mutable_column_view merged_view = p_merged_col->mutable_view();
+
+    //initialize null_mask to all valid:
+    //
+    //Note: this initialization in conjunction with _conditionally_
+    //calling materialze_bitmask() below covers the case
+    //materialize_merged_bitmask_kernel<false, false>()
+    //which won't be called anymore (because of the _condition_ below)
+    //
+    cudf::set_null_mask(merged_view.null_mask(),
+                        merged_view.size(),
+                        ALL_VALID,
+                        stream_);
+
+    //to resolve view.data()'s types use: ElemenT
+    //
+    ElemenT const* p_d_lcol = lcol.data<ElemenT>();
+    ElemenT const* p_d_rcol = rcol.data<ElemenT>();
+        
+    auto exe_pol = rmm::exec_policy(stream_);
+    
+    //capture lcol, rcol
+    //and "gather" into merged_view.data()[indx_merged]
+    //from lcol or rcol, depending on side;
+    //
+    thrust::transform(exe_pol->on(stream_),
+                      dv_row_order_.begin(), dv_row_order_.end(),
+                      merged_view.begin<ElemenT>(),
+                      [p_d_lcol, p_d_rcol] __device__ (index_type const& index_pair){
+                       auto side = thrust::get<0>(index_pair);
+                       auto index = thrust::get<1>(index_pair);
+                       
+                       ElemenT val = (side == side::LEFT ? p_d_lcol[index] : p_d_rcol[index]);
+                       return val;
+                      }
+                     );
+
+    //cudaDeviceSynchronize();//? nope...these two could proceed concurrently
+
+
+    //CAVEAT: conditional call below is erroneous without
+    //set_null_mask() call (see TODO above):
+    //
+    if (lcol.has_nulls() || rcol.has_nulls())
+      //resolve null mask:
+      //
+      materialize_bitmask(lcol,
+                          rcol,
+                          merged_view,
+                          dv_row_order_.data().get(),
+                          stream_);
+                   
+    return p_merged_col;
```

Comment:
It doesn't look like the null count for the output is ever updated. Note that it is still producing the correct result because it is lazily evaluating the null count in `column::null_count`. However, it would be more performance to simply do `p_merge_col.set_null_count(lcol.null_count() + rcol.null_count());`

Reply:
Done (eliminating `materialize_merged_bitmask_kernel<false, false>()`).

Reply:
Not sure I get your point about lazy evaluation of nulls. `materialize_bit_mask()` call is taking care of the nulls. Are you saying the null _count_ doesn't (always) get updated and I can just do a `set_null_count()` (after or before `materialize_bit_mask()`)? Am I understanding this correctly?

Reply:
Done (set_null_count()).